DOCS-#6: Update and extend readme.md (#87)
modin-project/unidist: README.md

<h2 align="center">Unified Distributed Execution</h2>

<p align="center">
<a href="https://unidist.readthedocs.io/en/latest/?badge=latest"><img alt="" src="https://readthedocs.org/projects/unidist/badge/?version=latest" align="center"></a>
</p>

### What is unidist?

unidist is a framework that is intended to provide the unified API for distributed execution by supporting various performant execution backends. At the moment the following backends are supported under the hood:

* [Ray](https://docs.ray.io/en/master/index.html)
* [MPI](https://www.mpi-forum.org/)
* [Dask Distributed](https://distributed.dask.org/en/latest/)
* [Python Multiprocessing](https://docs.python.org/3/library/multiprocessing.html)

unidist is designed to work in a [task-based parallel](https://en.wikipedia.org/wiki/Task_parallelism) model.

Also, the framework provides a sequential ``Python`` backend, that can be used for debugging.

### Installation

Currently, unidist can be installed from sources using ``pip``:

```bash

> $ pip install git+https://github.com/modin-project/unidist

Reviewer: It would be nice to say that in this case only `Python`/`MultiProcessing` backends will work. Or you can add several lines-examples of installation different backend-deps sets like:
```suggestion
$ pip install git+https://github.com/modin-project/unidist  # `MultiProcessing` and `Python` are supported only
$ pip install git+https://github.com/modin-project/unidist#egg=unidist[all]  # all deps
$ pip install git+https://github.com/modin-project/unidist#egg=unidist[ray]  # ray deps
```

Author: fixed

<p align="center"><a href="https://unidist.readthedocs.io"><img width=77% alt="" src="https://github.com/modin-project/unidist/blob/master/docs/img/unidist_logo.png?raw=true"></a></p>
<h2 align="center">Unified Distributed Execution</h2>

<p align="center">
<a href="https://github.com/modin-project/unidist/actions"><img src="https://github.com/modin-project/unidist/workflows/master/badge.svg" align="center"></a>
<a href="https://unidist.readthedocs.io/en/latest/?badge=latest"><img alt="" src="https://readthedocs.org/projects/unidist/badge/?version=latest" align="center"></a>
</p>

### What is unidist?

> unidist is a framework that is intended to provide the unified API for distributed execution by supporting various performant execution backends and utilizing all of your CPUs cores. At the moment the following backends

Author: There is a note on task-parallel manner below. Also, we plan to support SPMD mode for MPI. Left it as is.

> unidist is a framework that is intended to provide the unified API for distributed execution by supporting various performant execution backends and utilizing all of your CPUs cores. At the moment the following backends

Reviewer 2: Then remove "utilizing all of your CPU cores" phrase - it really depends on the settings and a workload

> unidist is a framework that is intended to provide the unified API for distributed execution by supporting various performant execution backends and utilizing all of your CPUs cores. At the moment the following backends

Author: fixed

> are supported underneath the hood:

Author: I would argue what is more suitable here. I think that is `underneath` so left as is.

> are supported underneath the hood:

Reviewer 2: **underneath the hood** is just not the commonly used phrase compared with the proposed one, I doubt it is event exists as a phrase.

> are supported underneath the hood:

Author: To make sure `underneath the hood` is being used as well you can search for it in GitHub but I do not have strong preferences on this so let's change it.

> are supported underneath the hood:

Author: fixed

* [Ray](https://docs.ray.io/en/master/index.html)
* [MPI](https://www.mpi-forum.org/)
* [Dask Distributed](https://distributed.dask.org/en/latest/)
* [Python Multiprocessing](https://docs.python.org/3/library/multiprocessing.html)

unidist is designed to work in a [task-based parallel](https://en.wikipedia.org/wiki/Task_parallelism) model.

Also, the framework provides a sequential ``Python`` backend, that can be used for debugging.

> ### Instalation

Reviewer: ```suggestion
### Installation
```

Currently, unidist can be installed from sources using ``pip``:

```bash
$ pip install git+https://github.com/modin-project/unidist
```

> This will install unidist directly from the repo without you having to manually clone it! Please be aware that these changes have not made it into a release and may not be completely stable.

Reviewer 2: ```suggestion
This will install unidist directly from the repo without you having to manually clone it! Please be aware that the latest changes have not made it into a release and may not be completely stable.
```

#### Choosing an execution backend

There are several ways to choose the execution backend for distributing computations.
First, the recommended way is to use
[unidist CLI](https://unidist.readthedocs.io/en/latest/using_cli.html) options:

```bash
# Running the script with unidist on Ray backend
$ unidist script.py --backend ray
# Running the script with unidist on MPI backend
$ unidist script.py --backend mpi
```

Second, setting the environment variable:

```bash
# unidist will use Ray backend to distribute computations
export UNIDIST_BACKEND=ray
# unidist will use MPI backend to distribute computations
export UNIDIST_BACKEND=mpi
```

Third, using [config API](https://unidist.readthedocs.io/en/latest/flow/unidist/config.html) directly in your script:

```python
import unidist.config as cfg

> cfg.Backend.put("mpi") # unidist will use MPI backend to distribute computations

Reviewer: Need to add note, that MPI backend must be run via `unidist script.py -b mpi` always with why we do that.

Author: We should put this note in using_mpi section so for now changed it to Dask usage to simplify Readme. Once we have more stable MPI and more docs on it, we will be able to expose MPI in Readme.

Reviewer: ok, will see in https://github.com/modin-project/unidist/issues/85

import unidist.config as cfg
cfg.Backend.put("ray") # unidist will use Ray backend to distribute computations
```

Default execution backend for unidist is Ray.

#### Usage

unidist provides [CLI interface](https://unidist.readthedocs.io/en/latest/using_cli.html) to run python programs.

```python
import unidist
unidist.init() # Ray backend is used by default

@unidist.remote
def foo(x):
    return x * x

> # This will run `foo` in parallel in workers;
> # `refs` will contain object references to actual data

Reviewer 2: ```suggestion
# Transform a `foo` in a unidist task and schedule it for parallel execution on a pool of workers;
# `refs` will contain object references to actual data
```

Author: Transforming  `foo` in a unidist task happens earlier. Also, I think this is a bit cumbersome phrase. Left it as is.

Reviewer 2: Ok, but I can`t understand the "run ... parallel in workers" phrase. At least "in" is not correct here. *run .. on a pool of workers in parallel" better for me.

Author: fixed
